### PR TITLE
Synchronize access to responseObserver

### PR DESCRIPTION
**Pull request by sschepens**
Synchronize access to `responseObserver` to prevent races:

```
java.lang.IllegalStateException: call already closed
    at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
    at io.grpc.internal.ServerCallImpl.close(ServerCallImpl.java:172)
    at io.grpc.internal.ServerCallImpl.sendMessage(ServerCallImpl.java:142)
    at io.grpc.stub.ServerCalls$ServerCallStreamObserverImpl.onNext(ServerCalls.java:341)
    at io.envoyproxy.controlplane.server.DiscoveryRequestStreamObserver.send(DiscoveryRequestStreamObserver.java:133)
    at io.envoyproxy.controlplane.server.DiscoveryRequestStreamObserver.lambda$processRequest$3(DiscoveryRequestStreamObserver.java:109)
    at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)
    at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909)
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    at java.base/java.lang.Thread.run(Thread.java:834)
```

Related to #106

### CODE REVIEW

File: server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java [hunk 1]
```diff
@@ -228,10 +228,15 @@ public void onNext(DiscoveryRequest request) {
...
-                  .asRuntimeException());
+          synchronized (responseObserver) {
+            if (!isClosing) {
+              isClosing = true;
+              responseObserver.onError(
+                  Status.UNKNOWN
+                      .withDescription(String.format("[%d] type URL is required for ADS", streamId))
+                      .asRuntimeException());
+            }
+          }
```
Comment (jakubdyszkiewicz) [inline server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java]: Can we use
```
closeWithError(Status.UNKNOWN .withDescription...)
```
I see that `closeWithError` also cancels the stream, but this is an exceptional situation anyway.

File: server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java [hunk 2]
```diff
@@ -292,7 +297,12 @@ public void onError(Throwable t) {
 
       try {
         callbacks.forEach(cb -> cb.onStreamCloseWithError(streamId, defaultTypeUrl, t));
-        responseObserver.onError(Status.fromThrowable(t).asException());
+        synchronized (responseObserver) {
+          if (!isClosing) {
+            isClosing = true;
+            responseObserver.onError(Status.fromThrowable(t).asException());
+          }
+        }
```
Comment (jakubdyszkiewicz) [inline server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java]: Can we use `closeWithError` here?

File: server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java (hunk 1 above)
Comment (sschepens) [reply server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java]: sure

**Comment by sschepens** [comment]
I'd like to add that this is actually not the solution i'd like to have.
In the future i think it would be better for us to bind the whole stream and watch to an executor to provide serialization guarantees and avoid having to synchronize everything

**Pull request merged**